README.md
# plaintext-accounting


CHANGELOG.org
#+Title: Changelog for crypto-accounting

This changelog will be maintained with sections for releases, denoted by release
date and version, with any unreleased code in an "unreleased" section.

* Pre-release version 0.0.4
** DONE Added Trade model
** DONE Fix shortcut generation to account for @ vs =.
  :PROPERTIES:
  :ARCHIVE_TIME: 2018-10-04 Thu 21:44
  :ARCHIVE_FILE: ~/ssd/work/crypto-accounting/TODO.org
  :ARCHIVE_OLPATH: Active Task List
  :ARCHIVE_CATEGORY: TODO
  :ARCHIVE_TODO: DONE
  :ARCHIVE_ITAGS: crypto_acct
  :END:
[[file:test/loaders/ledger_loader.test.js::t.is(results%5B0%5D.entries%5B1%5D.getFullShortcut(),%20'0.96555%20BTC%20Assets:Exchanges:Coinbase%20@%20200%20USD');][Test to modify]]


* Pre-release version 0.0.3
** DONE Corrected fees to be debits to expenses (not credit)
** DONE Settle capital gains in lots
** DONE Add a loader test with Fees
** DONE Make a gains test that includes fees for cap-gains

** DONE Possibly converting/loading from Ledger.dat is reversed!  Test and fix.
   CLOSED: [2018-09-03 Mon 20:25]
   - State "DONE"       from              [2018-09-03 Mon 20:25]
    See [[file:~/work/crypto-accounting/test/loaders/fixtures/ark.dat][file:~/work/crypto-accounting/test/loaders/fixtures/ark.dat]]
** DONE Work through a real ledger file, fixing parser to handle edge cases
   CLOSED: [2018-08-10 Fri 15:34]
   - State "DONE"       from              [2018-08-10 Fri 15:34]

** DONE Add day-buckets to wallet csv converter
   CLOSED: [2018-08-05 Sun 18:38]
   - State "DONE"       from              [2018-08-05 Sun 18:38]
** DONE Add basic CLI framework
   CLOSED: [2018-08-05 Sun 18:38]
   - State "DONE"       from              [2018-08-05 Sun 18:38]

** DONE Add build step
   CLOSED: [2018-07-29 Sun 16:46]
   - State "DONE"       from              [2018-07-29 Sun 16:46]

** DONE Change to require from import, for compatibility.
   CLOSED: [2018-07-28 Sat 07:47]
   - State "DONE"       from              [2018-07-28 Sat 07:47]

** DONE Create sanity/cleanliness report for Journal [4/4]
   CLOSED: [2018-07-27 Fri 14:49]
   - State "DONE"       from              [2018-07-27 Fri 14:49]
   - [X] currencies all defined
   - [X] accounts declared
   - [X] all transactions balanced
   - [X] Should return an object helping to find problems

** DONE Add docstrings
   CLOSED: [2018-07-27 Fri 12:55]
   - State "DONE"       from              [2018-07-27 Fri 12:55]
** DONE Add unrealized gains method
   CLOSED: [2018-07-26 Thu 18:35]
   - State "DONE"       from              [2018-07-26 Thu 18:35]
** DONE Add lot cap-gains
   CLOSED: [2018-07-25 Wed 19:03]
   - State "DONE"       from              [2018-07-25 Wed 19:03]
** DONE Build the price history converter
   CLOSED: [2018-07-15 Sun 21:41]
   - State "DONE"       from              [2018-07-15 Sun 21:41]
*** Possibly use a sorted-array to make for fast date searches
    https://github.com/javascript/sorted-array
** DONE Make a wallet-export converter
   CLOSED: [2018-07-13 Fri 16:17]
   - State "DONE"       from              [2018-07-13 Fri 16:17]

** DONE Add LIFO
   CLOSED: [2018-07-12 Thu 18:29]
   - State "DONE"       from "TODO"       [2018-07-12 Thu 18:29]

** DONE Find lots
   CLOSED: [2018-07-08 Sun 18:32]
   - State "DONE"       from "TODO"       [2018-07-08 Sun 18:32]
*** Add multiple entries (for bucketing)
*** Add disbursement entries
*** Add isClosed

** DONE Solve "where do assets sold to me by an exchange get credited?"
   CLOSED: [2018-07-04 Wed 12:59]
   - State "DONE"       from              [2018-07-04 Wed 12:59]
    Answer - designated equity account
** DONE Add virtual entries
   CLOSED: [2018-07-04 Wed 13:01]
   - State "DONE"       from              [2018-07-04 Wed 13:01]
** DONE Add balancing entries
   CLOSED: [2018-07-04 Wed 13:01]
   - State "DONE"       from              [2018-07-04 Wed 13:01]
** DONE Make getAccounts faster and easier to use
   CLOSED: [2018-07-04 Wed 13:01]
   - State "DONE"       from              [2018-07-04 Wed 13:01]

** DONE Add multiple aliases to accounts
   CLOSED: [2018-06-29 Fri 08:40]
   - State "DONE"       from              [2018-06-29 Fri 08:40]
** DONE Add party and address fields to Accounts
   CLOSED: [2018-06-29 Fri 08:40]
   - State "DONE"       from "TODO"       [2018-06-29 Fri 08:40]

** DONE Add alias support for accounts
   CLOSED: [2018-06-28 Thu 12:38]
   - State "DONE"       from              [2018-06-28 Thu 12:38]
** DONE Add total-by-currency
   CLOSED: [2018-06-27 Wed 18:31]
   - State "DONE"       from "TODO"       [2018-06-27 Wed 18:31]
** DONE Add total-by-filter to journal
   CLOSED: [2018-06-27 Wed 18:02]
   - State "DONE"       from "TODO"       [2018-06-27 Wed 18:02]
** DONE Remove carat from account shortcuts
   CLOSED: [2018-06-27 Wed 11:46]
   - State "DONE"       from              [2018-06-27 Wed 11:46]
** DONE Add total-to-date
   CLOSED: [2018-06-26 Tue 21:22]
   - State "DONE"       from              [2018-06-26 Tue 21:22]
** DONE Add journal total by account
   CLOSED: [2018-06-26 Tue 19:08]
   - State "DONE"       from "TODO"       [2018-06-26 Tue 19:08]
** DONE add account.getTotal()
   CLOSED: [2018-06-26 Tue 19:08]
   - State "DONE"       from              [2018-06-26 Tue 19:08]
** DONE add isBalanced to transaction.
   CLOSED: [2018-06-24 Sun 18:18]
   - State "DONE"       from "TODO"       [2018-06-24 Sun 18:18]
Must have equal # of debits and credits.
** DONE add auto-pairing of shortcuts
   CLOSED: [2018-06-24 Sun 17:50]
   - State "DONE"       from "TODO"       [2018-06-24 Sun 17:50]
ex: "1 ETH ^acct" would make a debit of 1 ETH to acct, and a credit of 1 ETH to tx.account.credit
"1 eth" would make a debit of 1 eth to tx.account.debit and a credit of 1 ETH to tx.account.credit

** DONE add automatic handling of transaction.account list for debit/credit.
   CLOSED: [2018-06-23 Sat 18:38]
   - State "DONE"       from "TODO"       [2018-06-23 Sat 18:38]
** DONE make sample file for accounts
   CLOSED: [2018-06-22 Fri 15:39]

scratch.js
var loader = require('./src/loaders/loader');
var journal = loader.loadJournalFromFilenameSync('journal_gains1.yaml', 'test/models/fixtures/');
var acct = journal.getAccount('cb');
//console.log(acct.toObject(true));

//console.log('debits', acct.getEntries('debit').map(e => e.toObject(true)));

// const lots = acct.getLots(journal.currencies);
// console.log('got');
// console.log('lots', lots);
// console.log('lot1', lots[0]);
// console.log('lot1', lots[0].toObject());

// var converter = require('./src/loaders/csv_converter');

// console.log(
//   converter.walletCsvToYamlSync(
//     '/home/bruce/Documents/Ledger/data/marco_2018-07-13.csv',
//     'MARCO',
//     'assets:crypto:wallet:MARCO',
//     'income:crypto:staking'
//   ));

// import PriceHistory from './src/models/pricehistory';

// const prices = [
//   '2018-06-16 ETH/USD 600',
//   '2018-06-17 BTC/USD 7000',
//   '2018-06-17 ETH/USD 650',
//   '2018-06-17 GIN/BTC 0.0001',
//   '2018-06-18 BTC/USD 6800',
//   '2018-06-18 ETH/USD 650',
//   '2018-06-18 GIN/BTC 0.00011',
//   '2018-06-19 ETH/USD 700',
//   '2018-06-20 ETH/USD 650',
// ];
// const history = new PriceHistory(prices);
// const price = history.findPrice('2018-06-17', 'GIN', 'USD', ['BTC']);
// console.log(price.toObject());

// import Transaction from './src/models/transaction';

// const transaction = new Transaction({
//   account: 'test',
//   utc: '2018-07-04',
//   entries: [
//     '-10 ETH @ 400 USD exchange',
//   ]
// });

// console.log(transaction.toObject());
// console.log('credits\n',transaction.getCredits().map(x => x.toObject()));
// console.log('debuts\n',transaction.getDebits().map(x => x.toObject()));

const Transaction = require('./src/models/transaction');

const tx = new Transaction({
    account: 'test',
    utc: '2018-07-04',
    trades: ['10 ETH @ 400 USD exchange']
  });

